docs/blueprints/reputation-store.md

```diff
@@ -0,0 +1,109 @@
+
+This document describes a solution for an interface to a central reputation store, which occasionally syncs with the `nodes` table. This allows us to transfer the workload of audits off of the `nodes` table to another store. 
+
+## Background
+
+The reason the `nodes` table has such poor performance is because we are doing a lot of row updates (e.g. audits, checkins) at the same time as we are trying to scan over the entire table (e.g. node selection, redash). Independently, row updates or scans work very well on Cockroach, but when combined, they can result in very undesirable performance.
+
+Having a separate store for reputation allows the audit service to make lots of updates for individual nodes, without needing to worry about bottlenecks caused by queries which need to do scans of the `nodes` table.
+
+Some fields are necessary in both the `nodes` table and the new reputation store. We will call these "contested" columns. Some examples are `disqualified` and `suspended`. Node selection needs to read them frequently, and audit needs to update them depending on audit outcomes.
+
+Through [denormalization](https://en.wikipedia.org/wiki/Denormalization), we can duplicate these fields in two separate stores so that the audit service can update the `nodes` table only when needed. The uncontested columns can be frequently updated by the audit service and the contested columns can be frequently read by node selection. The contested columns will only need to be updated by the audit service if one of them changes.
+
+The reputation store will be considered the source of truth for the contested values, as it will contain the most recently-updated values for reputation.
+
+## Design / Implementation
+
+### Reputation Package
+
+The reputation package will be responsible for maintaining the reputation store and updating the `nodes` table when necessary.
+
+It will contain a service with the following functionality. We can change this in the future if needed - this is just a starting point.
+
+```
+type Service struct {
+    overlay *overlay.Service
+    db      DB
+    config  Config
+}
+
+func (service *Service) ApplyAudit(nodeID storj.NodeID, overlay.AuditType) error
+func (service *Service) Get(nodeID storj.NodeID) (*Reputation, error)
+func (service *Service) TestingDisqualify(nodeID storj.NodeID) error
+func (service *Service) TestingSetState(state Reputation) error
+```
+
+Here are some type definitions for the above. We should also move types such as `AuditHistory` and `AuditType` out of the `overlay` package and into the `reputation` package. 
+
+```
+// these are the values that need to be updated in the nodes table if they change
+type statusChange struct {
+    Contained bool
+    Disqualified *time.Time
+    Suspended *time.Time
+    UnknownAuditSuspended *time.Time
+    OfflineSuspended *time.Time
+    VettedAt *time.Time
+}
+
+// these are all reputation values, including those in the statusChange struct
+type Reputation struct {
```

Comment:
Note: last_contact_success is updated by node check in and, ideally every single audit.
How do we reconcile this? Maybe last_contact_success should also be part of Reputation?

Reply:
last_contact_success should stay in the nodes table imo, and only be updated if it's more than say an hour ago

Reply:
last_contact_success should be updated by the contact service, so it should be fine if audit doesn't touch it - just as long as it keep track of audit histories

Reply:
Does this mean, audit is no longer going to update contact success field for nodes? Is that changing the current behavior?

Reply:
We can probably do it either way. If we want to update last contact success for audits, then we need to make sure that `contact` also updates it using the `reputation`store. But if we only care about `last_contact_success` from `contact`, we don't need to include that in the reputation store

Reply:
Yea, I think it maybe worth pointing it out in the design doc whether contact information is part of the reputation store

Reply:
Yingrong pointed this out: 
last_contact_success is also a part of "reputation" per se. It is used to determine whether a node should be selected or not.
What if we moved last_contact_success to the reputation store and rolled it into the "Reliable" column in nodes?

Reply:
I'm fine with moving it to the reputation store. We just need to make sure the contact service updates it there rather than via the overlay